### ISSUE TITLE
Test failure: StreamDoesNotStartIfTokenAlreadyCanceled

### ISSUE THREAD

**Issue by ryanbrandenburg** [aspnet/SignalR#2790]
This test [fails](http://aspnetci/viewLog.html?buildId=528685&buildTypeId=Releases_21Public_UbuntuUniverse) occasionally with the following error:
```
System.Exception : 1 error(s) logged.
Microsoft.AspNetCore.Http.Connections.Client.HttpConnection - TransportThrewExceptionOnStop - The transport threw an exception while stopping.
===================
System.Net.Http.HttpRequestException: Response status code does not indicate success: 404 (Not Found).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.SendUtils.<SendMessages>d__0.MoveNext() in /_/src/Microsoft.AspNetCore.Http.Connections.Client/Internal/SendUtils.cs:line 81
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.ServerSentEventsTransport.<ProcessAsync>d__18.MoveNext() in /_/src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs:line 105
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.ServerSentEventsTransport.<StopAsync>d__20.MoveNext() in /_/src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs:line 228
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.<DisposeAsyncCore>d__43.MoveNext() in /_/src/Microsoft.AspNetCore.Http.Connections.Client/HttpConnection.cs:line 256
===================
   at Microsoft.AspNetCore.SignalR.Tests.VerifyNoErrorsScope.Dispose() in /_/test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifyNoErrorsScope.cs:line 63
   at Microsoft.AspNetCore.SignalR.Tests.ServerLogScope.Dispose() in /_/test/Microsoft.AspNetCore.SignalR.Tests.Utils/ServerLogScope.cs:line 66
   at Microsoft.AspNetCore.SignalR.Client.FunctionalTests.HubConnectionTests.<StreamDoesNotStartIfTokenAlreadyCanceled>d__12.MoveNext() in /_/test/Microsoft.AspNetCore.SignalR.Client.FunctionalTests/HubConnectionTests.cs:line 424
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
------- Stdout: -------
| [2018-08-11T00:27:45] TestLifetime Information: Starting test StreamDoesNotStartIfTokenAlreadyCanceled_json_ServerSentEvents_default
| [2018-08-11T00:27:45] ServerLogScope Information: Server log scope started.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Waiting on Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:211).
| [2018-08-11T00:27:45] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Starting HubConnection.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Starting HttpConnection.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Establishing connection with server at 'http://127.0.0.1:45536/default'.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Http.Connections.Client.Internal.LoggingHttpMessageHandler Trace: Sending HTTP request POST 'http://127.0.0.1:45536/default/negotiate'.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets Debug: Connection id "0HLFV44RF39GQ" received FIN.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLFV44RF39GQ" disconnecting.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLFV44RF39H0" started.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets Debug: Connection id "0HLFV44RF39GQ" sending FIN.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request starting HTTP/1.1 POST http://127.0.0.1:45536/default/negotiate application/x-www-form-urlencoded 0
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler Debug: AuthenticationScheme: Bearer was not authenticated.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLFV44RF39GQ" stopped.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Routing.RouteBase Debug: Request successfully matched the route with name '(null)' and template '/default/negotiate'.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager Debug: New connection fpvZ5xSFDr6LzDdRZteRKA created.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher Debug: Sending negotiation response.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLFV44RF39H0" completed keep alive response.
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request finished in 1.7054ms 200 application/json
| [2018-08-11T00:27:45] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Established connection 'fpvZ5x
...
```
Other tests within that build may have failed with a similar message, but they are not listed here. Check the link above for more info.

This test failed on 2.1.

CC @muratg

This issue was made automatically. If there is a problem contact ryanbrandenburg.

**Comment by ryanbrandenburg** [comment]
Please use this workflow to address this flaky test issue, including checking applicable checkboxes and filling in the applicable "TODO" entries:

* Is this actually a flaky test?
  * No, this is a regular test failure, fix the test/product (TODO: Link to commit/PR)
  * Yes, proceed below...

* Is this test failure caused by product code flakiness? (Either this product, or another product this test depends on.)
  * [ ] File a bug against the product (TODO: Link to other bug)
  * Is it possible to change the test to avoid the flakiness?
    * Yes? Go to the "Change the test!" section.
    * No?
      * [ ] Disable the test (TODO: Link to PR/commit)
      * [ ] Wait for other bug to be resolved
      * [ ] Wait for us to get build that has the fix
      * [ ] Re-enable our test (TODO: Link to PR/commit)
      * [ ] Close this bug

* Is it that the test itself is flaky? This includes external transient problems (e.g. remote server problems, file system race condition, etc.)
  * Is there is a way to change our test to avoid this flakiness?
    * Yes? Change the test!
      * [ ] Change the test to avoid the flakiness, for example by using a different test strategy, or by adding retries w/ timeouts (TODO: Link to PR/commit)
      * [ ] Run the test 100 times locally as a sanity check.
      * [ ] Close this bug
    * No?
      * Is there any logging or extra information that we could add to make this more diagnosable when it happens again?
        * Yes?
            * [ ] Add the logging (TODO: Link to PR/commit)
        * No?
            * [ ] Delete the test because flaky tests are not useful (TODO: Link to PR/commit)

This comment was made automatically. If there is a problem contact ryanbrandenburg.

**Comment by BrennanConroy** [comment]
FYI, this issue is from 2.1 builds. We fixed the flaky test in 2.2 see https://github.com/aspnet/SignalR/pull/2739

We should consider porting the fix to 2.1

**Comment by natemcmaster** [comment]
Fixed in https://github.com/aspnet/SignalR/pull/2805
